File: packages/canonical-json-ts/src/path.ts
/**
 * Build a JSON-pointer (RFC 6901) path from a sequence of segments.
 *
 * Segments are escaped per RFC 6901: `~` → `~0`, `/` → `~1`.
 * The empty path (root) is represented as the empty string.
 */
export function joinPath(parent: string, segment: string | number): string {
  const escaped =
    typeof segment === 'number'
      ? String(segment)
      : segment.replace(/~/g, '~0').replace(/\//g, '~1');
  return `${parent}/${escaped}`;
}

File: packages/canonical-json-py/src/agentpay_canonical_json/path.py
"""
JSON-pointer (RFC 6901) path utilities.
"""


def join_path(parent: str, segment: str | int) -> str:
    """
    Build a JSON-pointer (RFC 6901) path from a sequence of segments.

    Segments are escaped per RFC 6901: `~` → `~0`, `/` → `~1`.
    The empty path (root) is represented as the empty string.
    """
    if isinstance(segment, int):
        escaped = str(segment)
    else:
        escaped = segment.replace("~", "~0").replace("/", "~1")
    return f"{parent}/{escaped}"
